bytecode-verificator/bytecode_verificator.sh

```diff
@@ -0,0 +1,479 @@
+            ;;
+          *) _err "Unknown constructor argument type '$arg_type', use --constructor-calldata instead" ;;
+        esac
+      done
+
+      for index in "${!compl_data[@]}"; do
+        encoded_data_length=$(bc <<< "${#constructor_calldata} / 2")
+        constructor_calldata=$(sed -E "s/$(endode_placeholder "$index")/$(printf "%064X\n" "$encoded_data_length")/" <<< "$constructor_calldata${compl_data[$index]}")
+      done
+    fi
+  fi
+
+  echo "Contract constructor encoded args: 0x$constructor_calldata"
+
+  echo "Deploying compiled contract to local fork"
+  contract_bytecode=$(cat ./build/"$contract_config_name".bin)
+  deployment_bytecode="0x$contract_bytecode$constructor_calldata"
+  deployer_account=$(_get_account "$local_rpc_url" 0)
+  local_contract_address=$(_deploy_contract "$local_rpc_url" "$deployer_account" "$deployment_bytecode")
+  echo "Done"
+}
+
+function compare_bytecode() {
+  local remote_code
+  local local_code
+  local etherscan_code
+
+  echo "Retrieving contract bytecode from local rpc (Ganache)"
+  local_code=$(_get_code $local_rpc_url "$local_contract_address")
+
+  echo -e "Retrieving contract bytecode from remote rpc ${remote_rpc_url}"
+  remote_code=$(_get_code "$remote_rpc_url" "$contract_config_address")
+
+  echo "Retrieving contract bytecode from etherscan"
+  etherscan_code=$(_get_code_etherscan "$contract_config_address")
+
+  echo "Replacing CBOR-encoded metadata"
+  # https://docs.soliditylang.org/en/v0.8.9/metadata.html#encoding-of-the-metadata-hash-in-the-bytecode
+  remote_code=$(sed -E 's/a264697066735822[0-9a-f]{68}//' <<< "$remote_code")
+  local_code=$(sed -E 's/a264697066735822[0-9a-f]{68}//' <<< "$local_code")
+  etherscan_code=$(sed -E 's/a264697066735822[0-9a-f]{68}//' <<< "$etherscan_code")
+
+  # https://docs.soliditylang.org/en/v0.4.24/metadata.html#encoding-of-the-metadata-hash-in-the-bytecode
+  remote_code=$(sed -E 's/a165627a7a72305820[0-9a-f]{68}//' <<< "$remote_code")
+  local_code=$(sed -E 's/a165627a7a72305820[0-9a-f]{68}//' <<< "$local_code")
+  etherscan_code=$(sed -E 's/a165627a7a72305820[0-9a-f]{68}//' <<< "$etherscan_code")
+
+  _print_checksum local_code
+  _print_checksum remote_code
+  _print_checksum etherscan_code
+
+  echo "Comparing remote and local bytecode"
+  [[ "$local_code" == "$remote_code" ]] ||  {
+    mkdir -p ./verificator_diffs
+    echo "$local_code" > ./verificator_diffs/"$contract"_local.bin
+    echo "$remote_code" > ./verificator_diffs/"$contract"_remote.bin
+    _err "local bytecode and remote bytecode is not equal. Bytecode saved in ./verificator_diffs/"
+  }
+  echo -e "${GREEN}Local bytecode matches with remote rpc${NC}"
+
+  echo "Comparing etherscan and local bytecode"
+  [[ "$local_code" == "$etherscan_code" ]] ||  { _err "local bytecode and etherscan bytecode is not equal"; }
+  echo -e "${GREEN}Local bytecode matches with etherscan${NC}"
+}
+
+# Internals
+
+_get_code() {
```

Comment:
Maybe add `local` variables to self-document the function and the others?

Reply:
Done